clustering/kubernetes.py

```diff
@@ -0,0 +1,449 @@
+    if info['status'] == 409:
+        name = data["metadata"].get("name", None)
+        info, body = api_request(module, url + "/" + name)
+        return False, body
+    return True, body
+
+
+def main():
+    module = AnsibleModule(
+        argument_spec=dict(
+            http_agent=dict(default=USER_AGENT),
+
+            url_username=dict(default="admin"),
+            url_password=dict(required=True, no_log=True),
+            force_basic_auth=dict(default="yes"),
+            validate_certs=dict(default=False, choices=BOOLEANS),
+            certificate_authority_data=dict(required=False),
+
+            # fake 'kubectl' commands
+            kubectl_api_versions=dict(required=False),
+            kubectl_cluster_info=dict(required=False),
+            kubectl_get=dict(required=False),
+            kubectl_namespace=dict(required=False, default="default"),
+            kubectl_version=dict(required=False),
+
+            # k8s API module variables
+            api_endpoint=dict(required=True),
+            file_reference=dict(required=False),
+            inline_data=dict(required=False),
+            state=dict(default="present",
+                       choices=["present", "post",
+                                "absent", "delete",
+                                "update", "put",
+                                "replace", "patch",
+                                "kubectl"])
+        )
+    )
+
+    decode_cert_data(module)
+
+    changed = False
+    data = module.params.get('inline_data', {})
+    if not data:
+        dfile = module.params.get('file_reference')
+        if dfile:
+            f = open(dfile, "r")
+            data = yaml.load(f)
+
+    endpoint = "https://" + module.params.get('api_endpoint')
+    url = endpoint
+
+    namespace = "default"
+    if data and 'metadata' in data:
+        namespace = data['metadata'].get('namespace', "default")
+        kind = data['kind'].lower()
+        url = endpoint + KIND_URL[kind]
+        url = url.replace("{namespace}", namespace)
+
+    # check for 'kubectl' commands
+    kubectl_api_versions = module.params.get('kubectl_api_versions')
+    kubectl_cluster_info = module.params.get('kubectl_cluster_info')
+    kubectl_get = module.params.get('kubectl_get')
+    kubectl_namespace = module.params.get('kubectl_namespace')
+    kubectl_version = module.params.get('kubectl_version')
+
+    state = module.params.get('state')
+    if state in ['present', 'post']:
+        changed, body = k8s_create_resource(module, url, data)
+        module.exit_json(changed=changed, api_response=body)
+
+    if state in ['absent', 'delete']:
+        changed, body = k8s_delete_resource(module, url, data)
+        module.exit_json(changed=changed, api_response=body)
+
+    if state in ['replace', 'put']:
+        changed, body = k8s_replace_resource(module, url, data)
+        module.exit_json(changed=changed, api_response=body)
+
+    if state in ['update', 'patch']:
```

Comment:
Updates should be done when state == 'present' and the current config in the cluster does not match.